src/security.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  BasePermissionValue,
+  EntityAttrPermissionValue,
+  EntityOperationType,
+  PermissionInfo,
+  PermissionType,
+  RoleInfo,
+  RoleType
+} from './model';
+
+/**
+ *
+ * Define which type of attribute render allowed for user
+ *
+ * @param entityName CUBA model entity
+ * @param attributeName
+ * @param perms - list of user permissions
+ * @param roles - list of user roles
+ * @return attribute could be not allowed to display (DENY), allowed for modification (MODIFY)
+ * or allowed in read only mode (VIEW).
+ */
+export function getAttributePermission(entityName: string,
+                                       attributeName: string,
+                                       perms: PermissionInfo[] | undefined,
+                                       roles: RoleInfo[] | undefined): EntityAttrPermissionValue {
+
+  if (!perms || !roles) return 'DENY';
+
+  if (hasRole(roles, RoleType.SUPER)) {
+    return 'MODIFY';
+  }
+
+  const perm = getMaxAllowedAttrPerm(entityName, attributeName, perms);
+
+  // return 'DENY' if no permission exist and user in STRICTLY_DENYING role
+  if (hasRole(roles, RoleType.STRICTLY_DENYING) && perm === null) {
+    return 'DENY';
+  }
+
+  return perm == null ? 'MODIFY' : perm.value as EntityAttrPermissionValue;
+}
+
+/**
+ * Define if operation (one of CRUD) on entity allowed or not for user
+ *
+ * @param entityName CUBA model entity
+ * @param operation - operation to be checked (CRUD)
+ * @param perms - list of user permissions
+ * @param roles - list of user roles
+ */
+export function isOperationAllowed(entityName: string,
+                                   operation: EntityOperationType,
+                                   perms: PermissionInfo[] | undefined,
+                                   roles: RoleInfo[] | undefined): boolean {
+
+  if (!perms || !roles) return false;
+
+  if (hasRole(roles, RoleType.SUPER)) return true;
+
+  const perm = getMaxAllowedOpPerm(entityName, operation, perms);
+
+  // readonly role not affect read operation
+  if (hasRole(roles, RoleType.READONLY) && operation !== 'read') {
+    // operation (except read) is disabled for readonly role if no perm is set
+    if (perm == null) return false;
+  }
+
+  if (hasRole(roles, RoleType.DENYING) || hasRole(roles, RoleType.STRICTLY_DENYING)) {
+    // operation is disabled for denying roles if no perm is set
+    if (perm == null) return false;
+  }
+
+  return perm == null || perm.value as BasePermissionValue !== 'DENY' ? true : false;
+}
+
+export function getMaxAllowedOpPerm(entityName: string,
+                                    operation: EntityOperationType,
+                                    perms: PermissionInfo[]): PermissionInfo | null {
+
+  const opFqn = `${entityName}:${operation}`;
+  let resultPerm = null;
+
+  perms
+    .filter(perm => perm.type === PermissionType.ENTITY_OP && perm.target === opFqn)
+    .forEach(perm => {
```

Comment:
Personally for such cases I'd prefer using `reduce` instead of mutating external variables.